src/hotspot/share/interpreter/bytecodeTracer.cpp

```diff
@@ -179,19 +176,17 @@ class BytecodePrinter {
 };
 
 #ifndef PRODUCT
-// We need a global instance to keep track of the states when the bytecodes
-// are executed. Access by multiple threads are controlled by ttyLocker.
-static BytecodePrinter _interpreter_printer;
+// We need a global instance to keep track of the method being printed so we can report that
+// the method has changed. If this method is redefined and removed, that's ok because the method passed in won't match, and
+// this will print that one.
+static Method* _current_method = nullptr;
```

Comment:
And if we are calling `trace_interpreter` from multiple threads they will each stomp on this shared global field. Sorry I can't see how some form of locking is not needed here.

Reply:
So actually this wasn't more racy than what was there in this respect.  You could change the global method before taking the ttyLock and for multiple threads, this would claim that the current method had changed.   The old code also had shared state for BytecodeTracer.  Now this uses a local instance, which can then print the bytecode at the given location.

Reply:
It would print the right name it would just print it again.  Say:
t1 printing M1 so prints the name
t1   aload
t1    astore
t2 printing M2 method name
t2  getfield
t1 < continues printing M1 but method name was changed to M2> prints M1 method name
t1 invokevirtual
...

The thread id is printed before each line.  Using this with multiple threads is not ideal for getting information clearly but that's not what people do.